refactor!: optional owner when getting notes

benesjan opened the pull request:
For the purposes of the `GlobalPrivateImmutable` which is being implemented in a PR up the stack the API of getting notes was insufficient because it forced me to specify owner which was not possible to do in `GlobalPrivateImmutable::view_note` because in unconstrained there is no concept of sender.

For this reason I make the owner optional in the API.

benesjan commented:
* **#18792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18815** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18815?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18815?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18782** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18782?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18721** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18721?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/history/note_inclusion/test.nr:
```diff
@@ -10,7 +10,7 @@ unconstrained fn succeeds_on_blocks_after_note_creation() {
         |context| {
             let header = context.anchor_block_header;
-            header.prove_note_inclusion(retrieved_note, test::NOTE_OWNER, test::NOTE_STORAGE_SLOT);
+            header.prove_note_inclusion(retrieved_note, test::NOTE_STORAGE_SLOT);
```
Now the owner is included in the retrieved_note because it needs to be there as it's important info to be received from the `get_notes` oracle.

nventuro replied:
The docs for retrieved note say that it's 'all necessary info to prove existence of a note'. The fact that we must also pass the slot to `prove_note_inclusion` shows that this is incorrect.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18815/commits/a1b6f79c1bb59b1358631edb39b95de67ac40d25

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_set.nr:
```diff
@@ -275,8 +275,11 @@ where
     {
-        let (retrieved_notes, note_hashes) =
-            get_notes(self.context, self.owner, self.storage_slot, options);
+        std::static_assert(options.owner.is_none(), "Owner cannot be set in pop_notes");
```
This is now a bit of an API regression as the input arg types allow setting of owner while it's incorrect to do here as it's defined by the state variable.

But this ugliness is fairly self-contained and should be always possible to figure out at comptime (Noir should see that we never attempt at setting it anywhere) and hence should not introduce constraints.

benesjan commented on noir-projects/aztec-nr/aztec/src/note/note_getter.nr:
```diff
@@ -78,7 +78,7 @@ fn check_notes_order<let N: u32>(
 pub fn get_note<Note>(
     context: &mut PrivateContext,
-    owner: AztecAddress,
+    owner: Option<AztecAddress>,
```
This function now looks a bit weird when called as we generally manual construct the option before calling the function. But it's mostly used state variables etc. and not by end user code so might be ok.

nventuro replied:
`RetrievedNote` is underconstrained here, you never checked that it is contains `owner` inside of it. This whole fn is now a bit hard to read and reason about - why is `storage_slot` outside of `RetrievedNote` but `owner` is inside of it? This may seem like a small thing, but it is ultimately what causes this bug.

nventuro replied:
This was not caught in tests because we have no equivalent for `constrain_get_note_internal`. Maybe we should, then it'd be more obvious. We in fact have no tests of constraining here.

benesjan replied:
Oh, this was a big fuckup. Agree with what you proposed and implemented it all in https://github.com/AztecProtocol/aztec-packages/pull/18815/commits/ffb2b826284a955bb7aaa738cc4aea254ff0bec6

Thanks

nventuro commented on noir-projects/aztec-nr/aztec/src/note/retrieved_note.nr:
```diff
@@ -7,6 +7,7 @@ use protocol_types::{address::AztecAddress, traits::{Deserialize, Packable, Seri
     pub contract_address: AztecAddress,
+    pub owner: AztecAddress,
     pub randomness: Field,
     pub metadata: NoteMetadata,
```
The docs lie, this is not all necessary info - storage slot is missing. Storage slot is now in the same category as randomness and owner, it's not special. The fact that `NewNote` contains storage slot but this one doesn't hints at that being possibly wrong.

benesjan replied:
Updated the docs in https://github.com/AztecProtocol/aztec-packages/pull/18815/commits/a1b6f79c1bb59b1358631edb39b95de67ac40d25.

>Storage slot is now in the same category as randomness and owner, it's not special.

I am not sure about this because storage slot is mandatory on the input of `get_notes` while everything else is not. This resulted in them being treated differently here.

I don't think there is a use case where we would want to not filter notes by storage slot so this is probably fine?

nventuro replied:
But isn't `RetrievedNote` the result of querying? I'm not sure why this being mandatory in the query is different.

nventuro replied:
Or is it because if the query takes the slot as a param, you feel it's repetitive to also include it in the result?

benesjan replied:
The storage slot is always defined when you request notes from the get_notes oracle and this is not the case with any other values.

Hence constrain-wise it makes sense to just use the constrained value you have from the input (since it's always defined) instead of needing to have there an assertion checking that the value in retrieved note matches input.

It's just one constraint though so the nicer API of having it in RetrievedNote might be worth it.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/execution_note_cache.ts:
```diff
@@ -173,12 +173,14 @@ export class ExecutionNoteCache {
     const notes = this.noteMap.get(contractAddress.toBigInt()) ?? [];
-    return notes.filter(n => n.note.owner.equals(owner) && n.note.storageSlot.equals(storageSlot)).map(n => n.note);
+    return notes
+      .filter(n => (owner === undefined || n.note.owner.equals(owner)) && n.note.storageSlot.equals(storageSlot))
```
```suggestion
      .filter(n => owner === undefined || n.note.owner.equals(owner))
      .filter(n => n.note.storageSlot.equals(storageSlot))
```

Nit, it's free to have more filtering steps (might even be faster), and I think this is much more readable.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/oracle/oracle.ts:
```diff
@@ -239,7 +239,8 @@ export class Oracle {
   }
 
   async utilityGetNotes(
-    [owner]: ACVMField[],
+    [ownerSome]: ACVMField[],
+    [ownerValue]: ACVMField[],
```
Shouldn't a change in the oracle API have triggered you requring to increment the oracle version number? That's not working correctly it seems.

benesjan replied:
Turns out the check is insufficient as we only check there oracle function names. Checking full signature is tricky but will try to figure it out.

In the meanwhile I bumped the version in https://github.com/AztecProtocol/aztec-packages/pull/18815/commits/d39d54dd091b59a62fdcb405e3860d38cc2e6805

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_getter/test.nr:
```diff
@@ -73,14 +69,10 @@ unconstrained fn constrain_get_notes_processes_many_notes() {
-            options,
-        );
+        let mut options = NoteGetterOptions::new();
+        options = options.set_owner(owner);
```
```suggestion
        let options = NoteGetterOptions::new().set_owner(owner);
```

No need to be overly verbose, the point of these methods is to allow chaining like this.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/commit/31e1d26cb1fc4fa2eed815ced2a43934dac6c644

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_getter/test.nr:
```diff
@@ -224,7 +238,9 @@ unconstrained fn constrain_get_notes_rejects_mismatched_selector() {
 
-        let _ = constrain_get_notes_internal(context, owner, storage_slot, opt_notes, options);
+        let mut options = options;
+        options = options.set_owner(owner);
```
???

Just do `NoteGetterOptions::new().select(...).set_owner(owner)`.

nventuro replied:
Might also want to rename `set_owner` to just `owner`.

benesjan replied:
Oh, this was quite retarded on my behalf. Thanks https://github.com/AztecProtocol/aztec-packages/pull/18815/commits/31e1d26cb1fc4fa2eed815ced2a43934dac6c644

Tackled there some this PR-unrelated occurrences to get rid of this everywhere.

> Might also want to rename set_owner to just owner.

We also have there `set_offset`, `set_limit` and `set_status` so if you think it's worth changing can do it in a separate PR.

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_getter/test.nr:
```diff
@@ -202,7 +194,29 @@ unconstrained fn constrain_get_notes_rejects_mismatched_address() {
+    let env = TestEnvironment::new();
+
+    env.private_context(|context| {
+        let different_owner = AztecAddress::from_field(100); // Different owner than the global owner
```
```suggestion
        let different_owner = AztecAddress::from_field(random());
```

I don't want to rely on the magical value not clashing into other magical valus

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_getter/test.nr:
```diff
@@ -202,7 +194,29 @@ unconstrained fn constrain_get_notes_rejects_mismatched_address() {
+        let note = MockNote::new(1)
+            .contract_address(context.this_address())
+            .owner(different_owner)
+            .build_retrieved_note(); // We're setting a different owner than what's in options
```
```suggestion
            .build_retrieved_note();
```

nventuro reviewed (commented):
The change itself seems good, but there's a couple concerning things, including a bug, a bad type which might be attributed to helping cause the bug, and oracle API CI checks not working.

benesjan commented on noir-projects/aztec-nr/aztec/src/oracle/version.nr:
```diff
@@ -30,7 +30,7 @@ mod test {
     }
 
-    #[test(should_fail_with = "Incompatible oracle version. TXE is using version '4', but got a request for '318183437'.")]
+    #[test(should_fail_with = "Incompatible oracle version. TXE is using version")]
```
Drop the end of the error message because having it there force updating it every oracle version update.

benesjan commented on noir-projects/aztec-nr/balance-set/src/balance_set.nr:
```diff
@@ -33,8 +33,9 @@ impl BalanceSet<UtilityContext> {
-        let notes = self.set.view_notes(options.set_offset(offset));
+        let options =
+            NoteViewerOptions::<UintNote, <UintNote as Packable>::N>::new().set_offset(offset);
+        let notes = self.set.view_notes(options);
```
Unrelated change which I did when I was going through the options and making them not mutable.

nventuro reviewed (approved):
My only remaining comment is re. `RetrievedNote`. It feels very natural to me to have it contain the slot, at which point it does provide all the data it needs to do what it is meant to do. Without the slot it is a bit of a useless value. Can you expand on why it feels wrong to you to include the slot there?

mverzilli commented on yarn-project/pxe/src/contract_function_simulator/execution_data_provider.ts:
```diff
@@ -76,15 +75,15 @@ export interface ExecutionDataProvider {
    *
    * @param contractAddress - The contract address of the notes.
-   * @param owner - The owner of the notes.
+   * @param owner - The owner of the notes. If undefined, returns notes for all owners.
```
Should we say: "If undefined, returns notes for all owners known to this PXE"?

"All owners" makes it feel like it's public

benesjan commented:
[The failure log](http://ci.aztec-labs.com/a00fef9cc49040dd) fails to open for me. Guess CI is broken so I am just aping this back in.

benesjan committed:
refactor!: optional owner when getting notes

For the purposes of the `GlobalPrivateImmutable` which is being implemented in a PR up the stack the API of getting notes was insufficient because it forced me to specify owner which was not possible to do in `GlobalPrivateImmutable::view_note` because in unconstrained there is no concept of sender.

For this reason I make the owner optional in the API.
